pallets/dapp-staking-v3/src/types.rs

```diff
@@ -170,108 +337,245 @@ where
+        amount: Balance,
+        era: EraNumber,
+    ) -> Result<(), SparseBoundedError> {
+        self.locked.add_amount(amount, era)
+    }
+
+    /// Subtracts the specified amount of the total locked amount, if possible.
+    ///
+    /// If entry for the specified era already exists, it's updated.
+    ///
+    /// If entry for the specified era doesn't exist, it's created and insertion is attempted.
+    /// In case vector has no more capacity, error is returned, and whole operation is a noop.
+    pub fn subtract_lock_amount(
+        &mut self,
+        amount: Balance,
+        era: EraNumber,
+    ) -> Result<(), SparseBoundedError> {
+        self.locked.subtract_amount(amount, era)
+    }
+
+    /// Adds the specified amount to the unlocking chunks.
+    ///
+    /// If entry for the specified block already exists, it's updated.
+    ///
+    /// If entry for the specified block doesn't exist, it's created and insertion is attempted.
+    /// In case vector has no more capacity, error is returned, and whole operation is a noop.
+    pub fn add_unlocking_chunk(
+        &mut self,
+        amount: Balance,
+        unlock_block: BlockNumber,
+    ) -> Result<(), SparseBoundedError> {
         if amount.is_zero() {
             return Ok(());
         }
 
-        let mut locked_chunk = if let Some(&locked_chunk) = self.locked.last() {
-            locked_chunk
-        } else {
-            LockedChunk::default()
-        };
-
-        locked_chunk.amount.saturating_accrue(amount);
+        let idx = self
+            .unlocking
+            .binary_search_by(|chunk| chunk.unlock_block.cmp(&unlock_block));
 
-        if locked_chunk.era == era && !self.locked.is_empty() {
-            if let Some(last) = self.locked.last_mut() {
-                *last = locked_chunk;
+        match idx {
+            Ok(idx) => {
+                self.unlocking[idx].amount.saturating_accrue(amount);
+            }
+            Err(idx) => {
+                let new_unlocking_chunk = UnlockingChunk {
+                    amount,
+                    unlock_block,
+                };
+                self.unlocking
+                    .try_insert(idx, new_unlocking_chunk)
+                    .map_err(|_| SparseBoundedError::NoCapacity)?;
             }
-        } else {
-            locked_chunk.era = era;
-            self.locked.try_push(locked_chunk).map_err(|_| ())?;
         }
 
         Ok(())
     }
+
+    /// Amount available for unlocking.
+    pub fn unlockable_amount(&self, current_period: PeriodNumber) -> Balance {
+        self.active_locked_amount()
+            .saturating_sub(self.active_stake(current_period))
+    }
+
+    /// Claims all of the fully unlocked chunks, and returns the total claimable amount.
+    pub fn claim_unlocked(&mut self, current_block_number: BlockNumber) -> Balance {
+        let mut total_unlocked = Balance::zero();
+
+        // Probably faster approach than partitioning, summing & collecting.
```

Comment:
All good with the current impl, tho personally I would prefer to sacrifice a bit of performance for readability.

Reply:
Totally agree!

First I wrote it using approach with `iters`, but it was just...unreadable 😓

Reply:
`.partition` might help? Like:

```rust
pub fn claim_unlock(...) -> Balance {
    let (unlocked, unlocking): (Vec<_>, Vec<_>) = self.unlocking.iter().partition(|chunk| chunk.unlock_block <= current_block_number);
    // assign new unlocking
    // fold `unlocked`
}
```

Reply:
I did it like that, but after Rust formatter broke it down into many lines, it looked messy (to me). So I'd actually prefer to keep it like this.

Reply:
It should be fine with good test coverage.

Actually, for me the formatted code makes it easier to see the `partition` method and helps with readability.